src/index.ts

```
@@ -1,34 +1,58 @@
-class User {
+interface User {
   name: string
   favoriteColor: string
   age: number
-  id: number
-
-  constructor(name: string, favoriteColor: string, age: number, id: number) {
-    this.name = name
-    this.favoriteColor = favoriteColor
-    this.age = age
-    this.id = id
-  }
+  id?: number
 }
 
 class UserAPI {
-  users: Array<User>
+  private users: Array<User>
+  private currentId: number
 
   constructor() {
     this.users = []
+    this.currentId = 1
   }
 
   private assignId() {
-
+    this.currentId++
+    return this.currentId - 1
   }
 
   addUser(user: User) {
-
+    try {
+      if (!user.hasOwnProperty("id")) {
+        const existingUser = this.users.find(
+          x => x.name === user.name
+            && x.favoriteColor === user.favoriteColor
+            && x.age === user.age)
+        if (existingUser !== undefined) {
+          throw "ERROR: User with these properties already exists"
+        }
+        user.id = this.assignId()
+        this.users.push(user)
+        return user
+      } else {
+        throw "ERROR: Id incorrectly provided by input user"
+      }
+    }
+    catch (error) {
```

rather than return a string or a user object, throw an error if the user is not found (see javascript.info on how to throw errors)

---

make sure to intentionally induce the error in the console log tests below